### ISSUE TITLE
Bug: maschere e CT scan non coincidono

### ISSUE THREAD

**Issue by AndreaCodegoni** [TommyBsk/Tiny_model_4_CD#4]
Come dicevamo ieri, maschere e CT scan non coincidono. Riporta qui quello che trovi e i risultati che ottieni quando riesci a far combaciare le maschere e CT.

**Comment by TommyBsk** [comment]
Dovrei aver trovato il problema, che si torva sempre nella funzione di downsamlpe dei dati che ti riporto di seguito:

```def downsample_seg(data, factor=2):
    def reduce_func(blocks, axis):
        H, W, D = blocks.shape[:3]
        Y, X, Z = np.meshgrid(np.arange(W),np.arange(H),np.arange(D))
        inds = Z + X*D + Y*W*D
        blocks_off = blocks + (inds*num_classes).reshape(*inds.shape, 1, 1, 1)
        counts = np.bincount(blocks_off.ravel(), minlength=inds.size*num_classes)
        counts = counts.reshape(inds.size, num_classes)[:,1:]
        max_counts = counts.max(axis=1)
        res = (counts.argmax(axis=1) + 1)
        res[max_counts==0] = 0
        return res.reshape(*X.shape).astype(np.uint8)

    num_classes = data.max() + 1
    logging.info(f"Downsampling labels by a factor of {factor}.")
    return block_reduce(data, block_size=(factor, factor, factor), func=reduce_func)`


Avevamo usato questa funzione per effettuare il downsample delle segmentazione. L'idea era di effettuare un downsample dei voxel.
Non sono ancora riuscito a sistemare la funzione ma il problema dovrebbe essere qui.

**Comment by AndreaCodegoni** [comment]
Se pensi sia questo il problema puoi provare a disabilitare questa funzione nella data prepare e poi controllare quello che ti ha generato sovrapponendo immagini e maschere. 
Se poi le sovrapposizioni sono perfette lavori su quel pezzettino li ma almeno sei sicuro che il problema è solo li.

**Comment by TommyBsk** [comment]
Sisi, avevo già provato a farlo andare senza il downsample ed effettivamente in questo modo le maschere corrispondono.

**Comment by TommyBsk** [comment]
Ho provate a fare qualche train e test_ondata senza usare il downsample, nella fase di train le metriche sono alte: F1 circa 0.9 e meanIOU circa di 0.83. Facendo però predizioni sui dati di test i numeri si abbassano molto, questi sono i risultati:
`{'acc': 0.997221676946985, 'miou': 0.5666757818281393, 'mf1': 0.6191239904499691, 'iou_0': 0.9972204600181106, 'iou_1': 0.13613110363816788, 'F1_0': 0.9986082362558466, 'F1_1': 0.2396397446440917, 'precision_0': 0.9974343705863001, 'precision_1': 0.6713089413800758, 'recall_0': 0.9997849876812616, 'recall_1': 0.14585269308390675}`
Mi è venuto il dubbio che sia dovuto a una segmentazione non perfetta fatta da me, visto che sto usando immagini binarie ottenute come unione delle semgmentazioni a 3 classi.
Sto cercando di sistemarle tutte a mano e appeno finisco provo a ricaricarle sul cluster e fare qualche prova.

**Comment by AndreaCodegoni** [comment]
Va bene. Tuttavia 0.23 di F1 e 0.13 di IoU è un valore estremamente basso che non penso sia riconducibile a una cattiva segmentazione, soprattutto se in training le metriche sono alte. Per convincertene puoi anche chiamate il test_ondata sul training e vedere che segmentazioni ti produce.

Il problema è che stai ancora overfittando. Al momento vedo tre strade da percorrere:

- fare una cartella in cui salvare per ogni slice una immagine che affianca CT scan, maschera e CT con sopra la maschera. Se poi non le vuoi vedere tu la prossima volta che ci sentiamo le guardo io o me le scarico io e le guardo. Mi dispiace essere pedante ma mi devo convincere che il dato ora sia corretto. Come hai visto anche la scorsa volta avevi questi risultati in termini di metriche quindi meglio essere certi.
- se tu dei dati sei molto sicuro allora lato training ci sono diverse strategie che puoi provare per non overfittare: meno epoche, diminuire il learning rate, introdurre delle altre augmentation, tenere la backbone freezzata ed allenare le restanti parti della rete...
- se vuoi ri labellizzare i dati vedi tu, e magari riesci a labializzare anche quelle nuove. tuttavia la terrei come strada da percorrere dopo aver provato le due descritte sopra o da fare mentre aspetti i risultati dei training.

Buona serata

**Comment by TommyBsk** [comment]
Perfetto, in realtà oggi ho create uno script che per tutte le immagini di test crea un immagine affiancando CT, gt e predizione.
Posso farlo anche per le immagini di training e validation se dovesse servire.
Per ora dalle immagini che ho guardato i dati e le segmentazioni mi sembrano corrette.

**Comment by AndreaCodegoni** [comment]
Le salverei tutte, poi ci diamo un occhio insieme

**Comment by TommyBsk** [comment]
Va bene, allora provo così.

**Comment by TommyBsk** [comment]
Sto facendo un train per salvare tutte le immagini di train e validation con le rispettive maschere e predizioni. 
Ho fatto solo 8 epoche perché la parte di salvataggio richiede un po' di tempo va bene lo stesso o sarebbe meglio salvare più epoche?
